Source/Parallelization/WarpXComm.cpp

```diff
@@ -871,33 +873,78 @@ WarpX::SyncCurrent (
-        J_cp[lev][2]->setVal(0.0);
+        for (int lev = finest_level; lev >= 0; --lev)
+        {
+            const int ncomp = J_fp[lev][idim]->nComp();
+            auto const& period = Geom(lev).periodicity();
 
-        const IntVect& refinement_ratio = refRatio(lev-1);
+            if (lev < finest_level)
+            {
+                // On a coarse level, the data in mf_comm comes from the
+                // coarse patch of the fine level. They are unfiltered and uncommunicated.
+                // We need to add it to the fine patch of the current level.
+                MultiFab fine_lev_cp(J_fp[lev][idim]->boxArray(),
+                                     J_fp[lev][idim]->DistributionMap(),
+                                     ncomp, 0);
+                fine_lev_cp.setVal(0.0);
+                fine_lev_cp.ParallelAdd(*mf_comm, 0, 0, ncomp, mf_comm->nGrowVect(),
+                                        IntVect(0), period);
+                // We now need to create a mask to fix the double counting.
+                auto owner_mask = amrex::OwnerMask(fine_lev_cp, period);
+                auto const& mma = owner_mask->const_arrays();
+                auto const& sma = fine_lev_cp.const_arrays();
+                auto const& dma = J_fp[lev][idim]->arrays();
+                amrex::ParallelFor(fine_lev_cp, IntVect(0), ncomp,
+                [=] AMREX_GPU_DEVICE (int bno, int i, int j, int k, int n)
+                {
+                    if (mma[bno](i,j,k) && sma[bno](i,j,k,n) != 0.0_rt) {
+                        dma[bno](i,j,k,n) += sma[bno](i,j,k,n);
+                    }
+                });
+                // Now it's safe to apply filter and sumboundary on J_cp
+                if (use_filter)
+                {
+                    ApplyFilterJ(J_cp, lev+1, idim);
+                }
+                SumBoundaryJ(J_cp, lev+1, idim, period);
+            }
 
-        std::array<const MultiFab*,3> fine { J_fp[lev][0].get(),
-                                             J_fp[lev][1].get(),
-                                             J_fp[lev][2].get() };
-        std::array<      MultiFab*,3> crse { J_cp[lev][0].get(),
-                                             J_cp[lev][1].get(),
-                                             J_cp[lev][2].get() };
-        ablastr::coarsen::average::Coarsen(*crse[0], *fine[0], refinement_ratio );
-        ablastr::coarsen::average::Coarsen(*crse[1], *fine[1], refinement_ratio );
-        ablastr::coarsen::average::Coarsen(*crse[2], *fine[2], refinement_ratio );
-    }
+            if (lev > 0)
+            {
+                // On a fine level, we need to coarsen the current onto the
+                // coarse level. This needs to be done before filtering because
+                // filtering depends on the level. This is also done before any
+                // same-level communication because it's easier this way to
+                // avoid double counting.
```

Comment:
As above: double counting of which points is referred to? Between levels or on the same level between neighboring boxes that share ownership of the same index?
```suggestion
                // avoid double counting of ...
```